include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -378,12 +394,71 @@ class pool_memory_resource final : public device_memory_resource {
 
   // map of [stream_id, free_list] pairs
   // stream stream_id must be synced before allocating from this list to a different stream
-  std::map<cudaStream_t, free_list> stream_free_blocks_;
+  // std::map<cudaStream_t, free_list> stream_free_blocks_;
+  std::map<cudaEvent_t, free_list> stream_free_blocks_;
 
   std::set<block, rmm::mr::detail::compare_blocks<block>> allocated_blocks_;
 
   // blocks allocated from upstream: so they can be easily freed
   std::vector<block> upstream_blocks_;
+
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+  /**
+   * @brief RAII wrapper for a CUDA event
+   */
+  struct cuda_event {
+    cuda_event()
+    {
+      auto result = cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
+      assert(cudaSuccess == result);
+    }
+    ~cuda_event()
+    {
+      auto result = cudaEventDestroy(event);
+      assert(cudaSuccess == result);
+    }
+
+    cudaEvent_t event;
+  };
+#endif
+
+  cudaEvent_t get_event(cudaStream_t stream)
+  {
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    if (stream == 0) {
```

Comment:
I think we need to do something other than check against `0` for the default stream. Someone could pass `cudaStreamPerThread` explicitly, which is different from `0`.

Reply:
Yeah, thought about that. Need to check for either 0 or cudaStreamPerThread, good point.

Reply:
Fixed.